lib/Dialect/FIRRTL/Transforms/InnerSymbolDCE.cpp

```diff
@@ -0,0 +1,122 @@
+//===- InnerSymbolDCE.cpp - Delete Unused Inner Symbols----------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//===----------------------------------------------------------------------===//
+//
+// This pass removes inner symbols which have no uses.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/SubElementInterfaces.h"
+#include "mlir/IR/Threading.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "firrtl-inner-symbol-dce"
+
+using namespace mlir;
+using namespace circt;
+using namespace firrtl;
+using namespace hw;
+
+struct InnerSymbolDCEPass : public InnerSymbolDCEBase<InnerSymbolDCEPass> {
+  void runOnOperation() override;
+
+private:
+  void findInnerRefs(Attribute attr);
+  void insertInnerRef(InnerRefAttr innerRef);
+  void removeInnerSyms(FModuleOp op);
+
+  DenseSet<std::pair<StringAttr, StringAttr>> innerRefs;
+};
+
+/// Find all InnerRefAttrs inside a given Attribute.
+void InnerSymbolDCEPass::findInnerRefs(Attribute attr) {
+  // Check if this Attribute is an InnerRefAttr.
+  if (auto innerRef = dyn_cast<InnerRefAttr>(attr)) {
+    insertInnerRef(innerRef);
+    return;
+  }
+
+  // Check if any sub-Attributes are InnerRefAttrs.
+  if (auto subElementAttr = dyn_cast<SubElementAttrInterface>(attr))
+    subElementAttr.walkSubAttrs([&](Attribute subAttr) {
+      if (auto innerRef = dyn_cast<InnerRefAttr>(subAttr))
+        insertInnerRef(innerRef);
+    });
+}
+
+/// Add an InnerRefAttr to the set of all InnerRefAttrs.
+void InnerSymbolDCEPass::insertInnerRef(InnerRefAttr innerRef) {
+  StringAttr moduleName = innerRef.getModule();
+  StringAttr symName = innerRef.getName();
+
+  auto [iter, inserted] = innerRefs.insert({moduleName, symName});
+  if (!inserted)
+    return;
+
+  ++numSymbolsFound;
+
+  LLVM_DEBUG(llvm::dbgs() << DEBUG_TYPE << ": found " << moduleName
+                          << "::" << symName << '\n';);
+}
+
+/// Remove all InnerSymAttrs within the FModuleOp that are dead code.
+void InnerSymbolDCEPass::removeInnerSyms(FModuleOp moduleOp) {
+  // Walk all ops in the FModuleOp.
+  moduleOp.walk([&](Operation *op) {
+    // Check if the op has an InnerSymAttr.
+    auto innerSym = op->getAttrOfType<InnerSymAttr>("inner_sym");
```

Comment:
Generally we prefer using InnerSymOpInterface to grab inner symbols on operations, idea being attribute name is not sufficient to ensure intent and that the name is an implementation detail (subject to change).

Leaning on IST might be useful here, as it has `walkSymbols` that takes a callback and returns InnerSymTargets (representing inner symbols in their various forms: ops, ports, and fields within those).

That would help address #4391 and also fields which I believe this mishandles unfortunately: they're not considered for pruning (lost opportunity) but since they're stored within the `InnerSymAttr`  any references to them should keep the `InnerSymAttr` alive...

InnerSymAttr does have a `walkSymbols` method too which could be used here to check if any of the callback names are used in discovered InnerRef's?

Reply:
That's great, I didn't even realize we had that InnerSymOpInterface and InnerSymbolTable functionality. Those should definitely be used here.